src/main/java/nextstep/subway/domain/SubwayMap.java

```diff
@@ -17,42 +20,52 @@ public SubwayMap(List<Line> lines, PathType pathType) {
-                .forEach(it -> {
-                    SectionEdge sectionEdge = SectionEdge.of(it);
-                    graph.addEdge(it.getUpStation(), it.getDownStation(), sectionEdge);
-                    graph.setEdgeWeight(sectionEdge, pathType.weight(it));
-                });
-
-        // 지하철 역의 연결 정보(간선)을 등록
-        lines.stream()
-                .flatMap(it -> it.getSections().stream())
-                .map(it -> new Section(it.getLine(), it.getDownStation(), it.getUpStation(), it.getDistance(), it.getDuration()))
-                .forEach(it -> {
-                    SectionEdge sectionEdge = SectionEdge.of(it);
-                    graph.addEdge(it.getUpStation(), it.getDownStation(), sectionEdge);
-                    graph.setEdgeWeight(sectionEdge, pathType.weight(it));
-                });
+        addVertex(graph);
+        addEdge(graph);
 
-        // 다익스트라 최단 경로 찾기
-        DijkstraShortestPath<Station, SectionEdge> dijkstraShortestPath = new DijkstraShortestPath<>(graph);
-        GraphPath<Station, SectionEdge> result = dijkstraShortestPath.getPath(source, target);
+        GraphPath<Station, SectionEdge> result = findShortedPath(source, target, graph);
 
         List<Section> sections = result.getEdgeList().stream()
                 .map(SectionEdge::getSection)
                 .collect(Collectors.toList());
 
         return new Path(new Sections(sections));
     }
+
+    private GraphPath<Station, SectionEdge> findShortedPath(Station source, Station target, WeightedGraph<Station, SectionEdge> graph) {
+        // 다익스트라 최단 경로 찾기
+        DijkstraShortestPath<Station, SectionEdge> dijkstraShortestPath = new DijkstraShortestPath<>(graph);
+        return dijkstraShortestPath.getPath(source, target);
+    }
+
+    private void addEdge(WeightedGraph<Station, SectionEdge> graph) {
+        // 지하철 역의 연결 정보(간선)을 등록 (역방향 포함)
+        List<Section> sections =
+                Stream.concat(
+                        flatMap(lines, line -> line.getSections().stream()).stream(),
+                        flatMap(lines, line -> line.getOppositeSections().stream()).stream()
+                ).collect(Collectors.toList());
+
+        sections.forEach(section -> addEdge(graph, section));
+    }
+
+    private void addEdge(WeightedGraph<Station, SectionEdge> graph, Section it) {
+        SectionEdge sectionEdge = SectionEdge.of(it);
+        graph.addEdge(it.getUpStation(), it.getDownStation(), sectionEdge);
+        graph.setEdgeWeight(sectionEdge, pathType.weight(it));
+    }
+
+    private void addVertex(WeightedGraph<Station, SectionEdge> graph) {
+        // 지하철 역(정점)을 등록
+        List<Station> stations = flatMap(lines, line -> line.getStations().stream());
+        stations.forEach(graph::addVertex);
+    }
+
+    private <T, R> List<R> flatMap(List<T> list, Function<T, Stream<R>> function) {
+        return list.stream()
+                .flatMap(function)
+                .collect(Collectors.toList());
```

Comment:
```suggestion
        return list.stream()
                .flatMap(function);
```

필요하면 이렇게 하는게 어떨까요? 🤔 
stream api 에서 flatMap 반환 타입이 stream 이라서 stream 을 반환하길 기대하게 만드는 메서드명인 것 같아요 😅

Reply:
오호.. 훨씬 좋은 것 같습니다!